Source/Core/CameraLimiter.js

```diff
@@ -0,0 +1,132 @@
+     */
+    function CameraLimiter(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        /**
+         * The BoundingObject. Used to restrict camera positions of the {@link Camera} holding this CameraLimiter.
+         * @type {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox}
+         * @default undefined
+         */
+        this.boundingObject = options.boundingObject;
+
+        /**
+         * The minimum orientation values.
+         * @type {HeadingPitchRoll}
+         */
+        this.minHeadingPitchRoll = HeadingPitchRoll.clone(options.minimumHeadingPitchRoll);
+
+        /**
+         * The maximum orientation values.
+         * @type {HeadingPitchRoll}
+         */
+        this.maxHeadingPitchRoll = HeadingPitchRoll.clone(options.maximumHeadingPitchRoll);
+    }
+
+    /**
+     * @private
+     */
+    CameraLimiter.prototype.limitPosition = function(position, result) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.object('position', position);
+        Check.typeOf.object('result', result);
+        //>>includeEnd('debug');
+
+        if (defined(this.boundingObject)) {
+            result = this.boundingObject.projectedPoint(position, result);
+        }
+        return result;
+    };
+
+    var scratchMatrix4 = new Matrix4();
+    /**
+     * @private
+     */
+    CameraLimiter.prototype.limitOrientation = function(orientation, position, result) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.object('orientation', orientation);
+        Check.typeOf.object('result', result);
+        //>>includeEnd('debug');
+
+        if (!defined(this.minHeadingPitchRoll) && !defined(this.maxHeadingPitchRoll)) {
+            result = orientation;
+            return result;
+        }
+
+        var quat = Quaternion.fromHeadingPitchRoll(orientation);
+        var transform = Transforms.headingPitchRollToFixedFrame(position, orientation);
+        var quatResult = Quaternion.fromRotationMatrix(Matrix4.getRotation(transform, scratchMatrix4));
+        Quaternion.multiply(quatResult, quat, quatResult);
+        result = HeadingPitchRoll.fromQuaternion(quatResult);
+
+        var minHPR = this.minHeadingPitchRoll;
+        var maxHPR = this.maxHeadingPitchRoll;
+
+        if (defined(this.minHeadingPitchRoll)) {
+            result.heading = Math.max(minHPR.heading, orientation.heading);
+            result.pitch = Math.max(minHPR.pitch, orientation.pitch);
+            result.roll = Math.max(minHPR.roll, orientation.roll);
+        }
+        if (defined(this.maxHeadingPitchRoll)) {
+            result.heading = Math.min(maxHPR.heading, orientation.heading);
+            result.pitch = Math.min(maxHPR.pitch, orientation.pitch);
+            result.roll = Math.min(maxHPR.roll, orientation.roll);
+        }
+        return result;
```

Comment:
I'm still seeing some strange behavior for the Heading/Pitch/Roll limiting. Pitch is never limited, and heading only works when both min and max are set to `0.0` or `-/+infinity`.

I think the way this method should work would be to have the camera limiter take in the current orientation of the camera, and return the limited orientation. You could then take that value, and perform the matrix operations in the camera class itself in the `limitOrientation` function. Do you agree? We can then do all of the conversions from local to fixed frame there.

Reply:
just realized didnt convert back to global space for hpr after limiting the local version - also updated by putting the conversion in the camera class instead of in limithpr in the camera limiter class - up in next commit